### PR TITLE
rfc(feature): SDK Spans Aggregator

### PR DESCRIPTION
**Pull request by philipphofmann**
This RFC aims to find a strategy for batching spans together to avoid sending one envelope per span.

[Rendered RFC](https://github.com/getsentry/rfcs/blob/rfc/batch-span-ingestion/text/0126-sdk-spans-aggregator.md)

### CODE REVIEW

File: text/0126-batch-span-ingestion.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    And sends the envelope to Sentry
+    And resets the timeout
+    And clears the SpanBuffer
+
+Scenario: 1 span added app crashes
+    Given 1 span in the SpanBuffer
+    When the SDK detects a crash
+    Then the SDK does nothing with the SpanBuffer
+    And loses the spans in the SpanBuffer
+```
```
Comment (markushi) [inline text/0126-batch-span-ingestion.md:73]:
```suggestion
Scenario: Unfinished spans
    Given no span is in the SpanBuffer
    When the SDK starts a span but doesn't finish it
    Then the SpanBuffer is still empty
    
Scenario: Open span, many children
    Given no span is in the SpanBuffer
    When the SDK starts a span but doesn't finish it yet
    And starts+finishes 60 child spans
    When the span finishes
    Then the SpanBuffer contains 61 spans
    And sends the envelope to Sentry
```
```

**Review by markushi — COMMENTED** [review]
Nice one, I left a few comments!

File: text/0126-batch-span-ingestion.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+## What happened to carrier transactions?
+
+On Mobile, we initially agreed on using transactions as carriers for spans in the [RFC mobile transactions and spans](https://github.com/getsentry/rfcs/blob/760467b85dbf86bd8b2b88d2a81f1a258dc07a1d/text/0118-mobile-transactions-and-spans.md).
+We planned on implementing [carrier transactions](https://github.com/getsentry/team-mobile/issues/157), but then [reverted the RFC](https://github.com/getsentry/rfcs/pull/125) and decided to use span ingestion instead because it won’t take long until the [PR for it gets merged](https://github.com/getsentry/relay/pull/2620).
+
+# Options Considered
+
+## Option 1: Batch Span Ingestion <a name="option-1"></a>
+
+A strategy to achieve this is to keep a buffer of spans in memory and batch them together in envelopes. The buffer starts a timeout of x seconds when the SDK adds the first span. When the timeout exceeds, the buffer sends all spans no matter how many items it contains. The buffer also sends all items after the SDK captures y spans. When the buffer sends all spans, it resets its timeout and removes all spans in the buffer. The buffer handles both auto-instrumented and manual spans.
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply text/0126-batch-span-ingestion.md:32]: Yes, indeed. Only finished spans. I will update the RFC.

[the remark this answers is not in the document]
Comment (philipphofmann) [reply text/0126-batch-span-ingestion.md:32]: Yes, indeed. Only finished spans. I will update the RFC.

**Comment by AbhiPrasad** [comment]
One high level thing is if you had the chance to look at https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/batchprocessor/README.md?

Instead of span count, does it make sense to use byte size as a weight like we are doing for metrics aggregation? Should this be an implemented on the client or the transport?

**Comment by philipphofmann** [comment]
@AbhiPrasad, thanks for the link to OTel. I should also look for existing concepts in OTel for every feature we build. The Span Buffer is similar to the Batch Processor of OTel. I added the option to set the timeout to 0 to this RFC. Do you think we should align our Span Buffer more with the OTel Batch Processor? If I'm correct, we must send span children in the same envelope as their parents. So that's a requirement the OTel Batch Processor doesn't specify. Therefore, I think having our version of Span Buffer specific to spans is okay. Furthermore, I don't think we have needs for `send_batch_max_size`, `metadata_keys`, and `metadata_cardinality_limit`. Or am I missing something?

> Instead of span count, does it make sense to use byte size as a weight like we are doing for metrics aggregation?

How do you count the byte sizes of the objects? Do you serialize them and keep track of the byte size? I think that's a great idea because you don't care about how many spans you keep in memory, but more about how much memory they use. 

> Should this be an implemented on the client or the transport?

As the Span Buffer must only accept spans after sampling and any other processing that could strip out data and keeping track of the byte size is easier after serializing, I would vote for the transport. If you agree, I'm going to update the RFC.

**Comment by philipphofmann** [comment]
@AbhiPrasad, please have a look at https://github.com/getsentry/rfcs/pull/126#issuecomment-1857513610.

**Comment by AbhiPrasad** [comment]
> Do you think we should align our Span Buffer more with the OTel Batch Processor? If I'm correct, we must send span children in the same envelope as their parents. So that's a requirement the OTel Batch Processor doesn't specify. Therefore, I think having our version of Span Buffer specific to spans is okay

We don't necessarily have the requirement that span children are sent in the same envelope as their parents, otherwise we run into the same issue with transactions (buffering a bunch of stuff in memory).

I do think having our own specific spec for a span buffer is makes sense though, and that `send_batch_max_size`, `metadata_keys`, and `metadata_cardinality_limit` seem unnecessary for our use case atm.

> How do you count the byte sizes of the objects? Do you serialize them and keep track of the byte size? I think that's a great idea because you don't care about how many spans you keep in memory, but more about how much memory they use.

so I was actually wrong about this, we don't keep the exact byte size, but instead an approximate weight based on type of metric/what it contains. We can try a similar heuristic as well, just weight by different properties on the span.

> As the Span Buffer must only accept spans after sampling and any other processing that could strip out data and keeping track of the byte size is easier after serializing, I would vote for the transport. If you agree, I'm going to update the RFC.

Transport sounds like the correct decision to me!

**Comment by philipphofmann** [comment]
> We don't necessarily have the requirement that span children are sent in the same envelope as their parents,

I thought that was a requirement. @phacops, can you confirm that span children don't need to be sent in the same envelope as their parents?

> we don't keep the exact byte size, but instead an approximate weight based on type of metric/what it contains. We can try a similar heuristic as well, just weight by different properties on the span.

Okay, so I would suggest something like:
*

**Comment by AbhiPrasad** [comment]
> I think we could rename the SpanBuffer to SpansAggregator to align with metrics

That matches my mental model, let's do it!

> As the SDKs need to serialize the spans and don't need to change them in the span buffer, we could also serialize them and count the serialized bytes

Hmm I think this varies based on platform, not everyone can serialize, maybe we make this requirement flexible depending on sdk? So SDK can decide on weighting methodology?

**Review by markushi — APPROVED** [review]
Nice improvements, looks good to me!

**Comment by philipphofmann** [comment]
> Hmm I think this varies based on platform, not everyone can serialize, maybe we make this requirement flexible depending on sdk? So SDK can decide on weighting methodology?

@AbhiPrasad, I thought so too, I will update the RFC 👍.

**Pull request merged**